orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/ConditionalStanzaTransform.java

```diff
@@ -46,25 +47,18 @@ public void visitPipelineTemplate(PipelineTemplate pipelineTemplate) {
   }
 
   private <T extends Conditional> void trimConditionals(List<T> stages, PipelineTemplate template) {
-    if (stages == null) {
-      return;
-    }
-
-    for (T el : stages) {
-      if (el.getWhen() == null || el.getWhen().isEmpty()) {
-        continue;
-      }
-
-      RenderContext context = new DefaultRenderContext(templateConfiguration.getPipeline().getApplication(), template, trigger);
-      context.getVariables().putAll(templateConfiguration.getPipeline().getVariables());
-
-      for (String conditional : el.getWhen()) {
-        String rendered = renderer.render(conditional, context);
-        if (!Boolean.parseBoolean(rendered)) {
-          stages.remove(el);
-          return;
+    Optional.ofNullable(stages).ifPresent( allStages -> allStages
+      .stream()
+      .filter(stage -> stage.getWhen() != null && !stage.getWhen().isEmpty())
+      .forEach(stage -> {
+        RenderContext context = new DefaultRenderContext(templateConfiguration.getPipeline().getApplication(), template, trigger);
+        context.getVariables().putAll(templateConfiguration.getPipeline().getVariables());
+        for (String conditional : stage.getWhen()) {
+          String rendered = renderer.render(conditional, context);
+          if (!Boolean.parseBoolean(rendered)) {
+            stage.setConditional(true);
+          }
```

Comment:
This logic reads weird. I see why you wrote it the way you did, but looking at this conditional on its own (without seeing `ConfigStageInjectionTransform`), you'd think it was logically a bug. Definitely warrants a comment like you have below.